feat(CanGetSoulshard): added a new function to check if player can get soulshards

ptchblvck opened the pull request:
### Summary
> **⚠️ Important: ** Before Implementation make sure to add this to the `UltraHardcore.toc` file!

- the new file checks if the player is a warlock class and checks if the target will give the player experience points. If both conditions are met an icon is shown above the `CustomResourceBar`. This can be changed. Very important notice: I have not updated the `UltraHardcore.toc` import because it might cause merge conflicts. The file has to be added to the `.toc` file after implementation.

Massive shoutout to @tulhur for helping break down the xp levels and also for Missed Vixen and Munkee on the [Ultra Discord]() for suggesting this feature

### Screenshots / Video (optional)

On Target:
<img width="1232" height="1239" alt="image" src="https://github.com/user-attachments/assets/6015d18b-3f1a-4840-aade-49a56b4e0ec9" />

Tooltip:
<img width="836" height="351" alt="image" src="https://github.com/user-attachments/assets/602245a4-a4cf-4205-8a08-1304368ec138" />


### Testing Steps (in-game)


1. Test matrix:
- Have a warlock
- Target an enemy that is in xp range

2. Expected result:
   - frame should be visible

### Future Improvements

We could implement a toggle for this to be enabled/disabled. Haven't touched that yet, because I wasn't sure if needed.

ptchblvck committed:
feat(CanGetSoulshard): added a new function to check if player can get soulshards

the new file checks if the player is a warlock class and checks if the target will give the player experience points. If both conditions are met an icon is shown above the `CustomResourceBar`. This can be changed. Very important notice: I have not updated the `UltraHardcore.toc` import because it might cause merge conflicts. The file has to be added to the `.toc` file after implementation.

Massive shoutout to @tulhur for helping break down the xp levels and also for Missed Vixen and Munkee on the Ultra Discord for suggesting this feature

Tulhur commented:
@ptchblvck I think you could remove both, the check if the player is a warlock and the check if the player is at or above level 10 if you instead check if the player knows the spell drain soul (isSpellKnown=1120). This would not only reduce the amount of checks you have to do, but more importantly fix a bug you likely currently have (I haven't tested it yet!). I only read through the code and from what I can see, a level 10 warlock would get the icon even if the player hasn't learned drain soul yet, which is the only way to actually get soul shards. In addition to that, you could also further improve the whole thing by checking for bag space (no bag space = no soul shards) and you'd also need I think atleast 2 conditions to define the position of the icon.
1: Is the resource bar enabled? If not, where to put it?
2: Is "Buffs on resource bar" enabled?  If no, show it in the position the buffs would be and if yes, move it further up

Sorry for the long message, just trying to think of ways to break it in advance :)

ptchblvck committed:
feat(soulshard): this enables the missing option to enable or disable the feature

I added additional values to be able to set the `ShowSoulShard Indicator` for players in `SettingsOptionsTab.lua`, `LoadDBData.lua` and `PresetSections.lua`. I still have not added anything to the `UltraHardcore.toc` because I was afraid this PR might run into conflicts. Have a very close look at the comments. I kept in the class check (commented it out) and also added some context comments for the indicator icon. All is up fo debate and can be changed.

Thanks again for @tulhur for pre-implementation looking through the code and reminding me to check for the drain soul spell, rather than having a class check. Testing this will only be available for warlocks above level 10.

ptchblvck commented:
@Tulhur I added the changes for the spell check. This means testing can only be done by warlocks above level 10 (or if you change the code to some special debug that will return a true value in the `CanGainXP` function. I also think this function should probably be moved to a different file, but my brain won't allow it this late. Will try to do some refactoring these upcoming weeks.

To answer your questions the indicator was never anchored to the resource bar, so we're good here.
Also when the resource bar buffs are enabled it's not the end of the world, because the indicator can be moved to wherever you like it (position will be saved). Indicator can also be reset with a slash command either use `/uhcsi` or `/uhcresetsoulshardindicator`

Cheers :)

ptchblvck committed:
feat(soulshard): updated the function logic

I extracted the `PlayerKnowsDrainSoul` into a function to make the `CanGainXPFromTarget` a function that could be put into its separate file. Also adjusted the level difference referencing @tulhur's mention from the wowwiki. I'm unsure if these are the correct values, they definitely have to be tested in a real scenario. We can always come back to adjust the values, because I also extracted those into its separate function.

Yet again thanks to @tulhur for doing the testing and QA in this whole process, while not even implemented yet. We're looking at it from many angles.

ptchblvck committed:
feat(soulshard): updated the level range table

Thanks again to @tulhur for all the information provided regarding the level range. Updated the `playerLevelRange()` funtion in the `CanGetSoulShard.lua` file. This still needs some testing to ensure these values are correct.

ptchblvck commented:
I adjusted the functions and refactored the code to make it a little bit easier to read and also maybe enable extracting the `CanGainXPFromTarget()` function.
